VALIDATING.md

```diff
@@ -0,0 +1,218 @@
+    # It is null for DELETE operations.
+    "object": {"apiVersion":"autoscaling/v1","kind":"Scale",...},
+    # oldObject is the existing object.
+    # It is null for CREATE and CONNECT operations.
+    "oldObject": {"apiVersion":"autoscaling/v1","kind":"Scale",...},
+    # options contains the options for the operation being admitted, like meta.k8s.io/v1 CreateOptions, UpdateOptions, or DeleteOptions.
+    # It is null for CONNECT operations.
+    "options": {"apiVersion":"meta.k8s.io/v1","kind":"UpdateOptions",...},
+
+    # dryRun indicates the API request is running in dry run mode and will not be persisted.
+    # Webhooks with side effects should avoid actuating those side effects when dryRun is true.
+    # See http://k8s.io/docs/reference/using-api/api-concepts/#make-a-dry-run-request for more details.
+    "dryRun": false
+  }
+}
+}]
+```
+
+Response example:
+```
+cat <<EOF >> $VALIDATING_RESPONSE_PATH
+{"allowed": true}
+EOF
+```
+
+Customize error message:
+```
+cat <<EOF >> $VALIDATING_RESPONSE_PATH
+{"allowed": false,
+"status": {
+      "code": 403,
+      "message": "You cannot do this because it is Tuesday and your name starts with A"
+    }
+ }
+EOF
+```
+
+
+## HTTP server and Kubernetes configuration
+
+Shell-operator should create an HTTP endpoint with TLS support and register endpoints in the ValidatingWebhookConfiguration resource.
+
+There should be a Service for shell-operator (see [Availability](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#availability)).
+
+New options:
+
+```
+  --validating-webhook-server-cert="/validating-certs/server.crt"
+                                 A path to a server certificate for ValidatingWebhook. Can be set with $VALIDATING_WEBHOOK_SERVER_CERT.
+  --validating-webhook-server-key="/validating-certs/server-key.pem"
+                                 A path to a server private key for ValidatingWebhook. Can be set with $VALIDATING_WEBHOOK_SERVER_KEY.
+  --validating-webhook-CLUSTER-ca="/validating-certs/cluster-ca.pem"
+                                 A path to a cluster ca bundle for ValidatingWebhook. Can be set with $VALIDATING_WEBHOOK_CLUSTER_CA.
+  --validating-webhook-client-ca=VALIDATING-WEBHOOK-CLIENT-CA ...
+                                 A path to a server certificate for ValidatingWebhook. Can be set with $VALIDATING_WEBHOOK_CLIENT_CA.
+  --validating-webhook-service-name=VALIDATING-WEBHOOK-SERVICE-NAME ...
+                                 A name of a service in front of a shell-operator. Can be set with $VALIDATING_WEBHOOK_SERVICE_NAME.
+```
+
+## Hook execution
+
+1. The shell-operator starts WebhookManager service at the start after starting informers and successful Synchronization.
+
+2. WebhookManager service starts the HTTP server.
+
+3.  WebhookManager service recreates `ValidatingWebhookConfiguration` resource. Items in `kubernetesValidating` in hooks are expanded with `clientConfig` field that contains a `path` field.
```

Comment:
If I deploy three replicas of a shell-operator to provide high availability, what will happen? I will end up with three ValidatingWebhookConfiguration created by each instance? Is it ok?

Reply:
The documentation suggests to create a Service to balance between replicas and one ValidatingWebhookConfiguration is enough.
The solution seems simple: each replica creates `ValidatingWebhookConfiguration` resource and ignores "Already exists" error.

Reply:
Sounds good. Just wanted to make it clear :)